File: C++/TwoPointer/sortColors.cpp
/*
Solution：双指针，时间复杂度O(n)，空间复杂度O(1)
- 设置指针p, q分别指向0区的最后一个位置和2区的第一个位置；
- 遍历数组，如果当前值为0，与nums[++p] 进行互换；如果为2，与nums[--q] 互换；为1的话，不进行操作；
- 注意当nums[i]等于2时，没有进行i++的操作，因为后面的数值被换到前面去，要对这个换过去的数值重新检查；
*/
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int p = -1, q = nums.size();
        int i = 0;
        while(i < q){
            if(nums[i] == 0){
                nums[i] = nums[++p];
                nums[p] = 0;
                i++;
            }
            else if(nums[i] == 2){
                nums[i] = nums[--q];
                nums[q] = 2;
            }
            else    i++;
        }
    }
};

File: Leetcode高频题/sortColors.py
'''
@lsy 2019.11.16

双指针。
p0/p2 分别指向 0/2 所在的最后一个位置。
遍历数组，若当前数字是 0，则与 p0 位置数字进行交换。
注意：
  判断 i < pointer2，要用 <，不要用 !=
'''


class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        pointer0, pointer2 = 0, len(nums) - 1
        i = 0
        while i < len(nums):
            if nums[i] == 0 and i > pointer0:
                nums[i], nums[pointer0] = nums[pointer0], nums[i]
                pointer0 += 1
            elif nums[i] == 2 and i < pointer2:
                nums[i], nums[pointer2] = nums[pointer2], nums[i]
                pointer2 -= 1
            else:
                i += 1
